providers/vercel/request.go

```diff
@@ -0,0 +1,298 @@
+package vercel
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/StackExchange/dnscontrol/v4/pkg/printer"
+	vercelClient "github.com/vercel/terraform-provider-vercel/client"
+)
+
+type clientRequest struct {
+	ctx              context.Context
+	method           string
+	url              string
+	body             string
+	errorOnNoContent bool
+}
+
+func (cr *clientRequest) toHTTPRequest() (*http.Request, error) {
+	r, err := http.NewRequestWithContext(
+		cr.ctx,
+		cr.method,
+		cr.url,
+		strings.NewReader(cr.body),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// Use a custom user agent for dnscontrol
+	r.Header.Set("User-Agent", "dnscontrol https://github.com/StackExchange/dnscontrol/pull/3542")
+	if cr.body != "" {
+		r.Header.Set("Content-Type", "application/json")
+	}
+	return r, nil
+}
+
+// doRequest is a helper function for consistently requesting data from vercel.
+// It implements rate limiting and retries.
+func (c *vercelProvider) doRequest(req clientRequest, v interface{}, rl *rateLimiter) error {
+	// Use a default http client with timeout
+	httpClient := &http.Client{
+		Timeout: 5 * 60 * time.Second,
+	}
+
+	for {
+		r, err := req.toHTTPRequest()
+		if err != nil {
+			return err
+		}
+		r.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.apiToken))
+
+		if rl == nil {
+			panic("doRequest is expecting a rate limiter but got nil, please fire an issue")
+		}
+
+		rl.delayRequest()
+
+		resp, err := httpClient.Do(r)
+		if err != nil {
+			return fmt.Errorf("error doing http request: %w", err)
+		}
+
+		// Handle rate limiting and retries, 429 is handled here
+		retry, err := rl.handleResponse(resp)
+
+		if err != nil {
+			resp.Body.Close()
```

Comment:
you could use `defer resp.Body.Close()` if suitable, optional suggestion.

Reply:
I did update to use defer, not sure if it is ok though. cc @tlimoncelli what do you think?